pkg/registry/core/service/rest.go

```diff
@@ -350,6 +350,37 @@ func (rs *REST) Update(ctx genericapirequest.Context, name string, objInfo rest.
 		return nil, false, errors.NewConflict(api.Resource("services"), service.Namespace, fmt.Errorf("Service.Namespace does not match the provided context"))
 	}
 
+	// TODO: this should probably move to strategy.PrepareForCreate()
+	releaseServiceIP := false
+	defer func() {
+		if releaseServiceIP {
+			if helper.IsServiceIPSet(service) {
+				rs.serviceIPs.Release(net.ParseIP(service.Spec.ClusterIP))
+			}
+		}
+	}()
+
+	if oldService.Spec.Type == api.ServiceTypeExternalName && helper.IsServiceIPRequested(service) {
+		// Allocate next available.
+		ip, err := rs.serviceIPs.AllocateNext()
+		if err != nil {
+			// TODO: what error should be returned here?  It's not a
+			// field-level validation failure (the field is valid), and it's
+			// not really an internal error.
+			return nil, false, errors.NewInternalError(fmt.Errorf("failed to allocate a serviceIP: %v", err))
+		}
+		service.Spec.ClusterIP = ip.String()
+		releaseServiceIP = true
+	} else if oldService.Spec.Type == api.ServiceTypeExternalName && helper.IsServiceIPSet(service) {
```

Comment:
What if the type is still ExternalName ?  You can end up allocating an IP when you don't need one.

I'm afraid that the accumulation of features in Service makes this code really tricky to reason about.  What I'd really like to see here is something more explicit, maybe a matrix of X->Y types and callouts for each that are consistent with Create()